futures-await-async-macro/src/lib.rs

```diff
@@ -224,42 +239,79 @@ where F: FnOnce(&Type) -> proc_macro2::TokenStream
 
 #[proc_macro_attribute]
 pub fn async(attribute: TokenStream, function: TokenStream) -> TokenStream {
-    // Handle arguments to the #[async] attribute, if any
     let (boxed, send) = match &attribute.to_string() as &str {
-        "( boxed )" => (true, false),
-        "( boxed_send )" => (true, true),
+        "( anchored )" => (true, false),
+        "( anchored_send )" => (true, true),
         "" => (false, false),
-        _ => panic!("the #[async] attribute currently only takes `boxed` as an arg"),
+        _ => panic!("the #[async] attribute currently only takes `anchored` as an arg"),
     };
 
-    async_inner(boxed, function, quote_cs! { ::futures::__rt::gen }, |output| {
+    async_inner(boxed, true, function, quote_cs! { ::futures::__rt::gen_pinned }, |output, lifetimes| {
         // TODO: can we lift the restriction that `futures` must be at the root of
         //       the crate?
         let output_span = first_last(&output);
         let return_ty = if boxed && !send {
             quote_cs! {
-                ::futures::__rt::std::boxed::Box<::futures::Future<
+                ::futures::__rt::PinBox<::futures::__rt::Future<
                     Item = <! as ::futures::__rt::IsResult>::Ok,
                     Error = <! as ::futures::__rt::IsResult>::Err,
                 >>
             }
         } else if boxed && send {
             quote_cs! {
-                ::futures::__rt::std::boxed::Box<::futures::Future<
+                ::futures::__rt::PinBox<::futures::__rt::Future<
                     Item = <! as ::futures::__rt::IsResult>::Ok,
                     Error = <! as ::futures::__rt::IsResult>::Err,
                 > + Send>
             }
+        } else {
+            quote_cs! {
+                impl ::futures::__rt::MyStableFuture<!> + #( #lifetimes + )*
+            }
+        };
+        let return_ty = respan(return_ty.into(), &output_span);
+        replace_bang(return_ty, &output)
+    })
+}
+
+#[proc_macro_attribute]
+pub fn async_move(attribute: TokenStream, function: TokenStream) -> TokenStream {
+    // Handle arguments to the #[async_move] attribute, if any
+    let (boxed, send) = match &attribute.to_string() as &str {
+        "( boxed )" => (true, false),
+        "( boxed_send )" => (true, true),
+        "" => (false, false),
+        _ => panic!("the #[async_move] attribute currently only takes `boxed` as an arg"),
+    };
+
+    async_inner(boxed, false, function, quote_cs! { ::futures::__rt::gen_move }, |output, lifetimes| {
+        // TODO: can we lift the restriction that `futures` must be at the root of
+        //       the crate?
+        let output_span = first_last(&output);
+        let return_ty = if boxed && !send {
+            quote_cs! {
+                ::futures::__rt::std::boxed::Box<::futures::__rt::Future<
+                    Item = <! as ::futures::__rt::IsResult>::Ok,
```

Comment:
I don't understand what's going on here-- why is `!` type involved?

Reply:
I see the `replace_bang` function, but I don't get what's going on there.

Reply:
copied for existing code, i can't help

Reply:
Ping @alexcrichton

Reply:
Oh it's a really bad hack to work with spans. The general idea is that I want to rewrite the spans of the tokens generated in this particular macro invocation but I don't want to overwrite user-provided tokens. The hack here was to create a macro with weird tokens (like `!`) and then the spans are rewritten. Afterwards the users's tokens are spliced into where the `!` is located.

In that the `!` here is just a placeholder, it doesn't make its way to the end

Reply:
Okay. We should try and address this in a followup (esp. `!` is now a type), but @withoutboats i'm fine w/ leaving as-is for now. Thanks for explaining @alexcrichton!